### PR TITLE
All AI Lawsets are rebalanced, can be researched, appear in config, and random spawners for AI upload.

### LINKED ISSUE #43197: AI law uploading's fucked
Issue reported from  Round ID: 104106 (/tg/Station Sybil [ENGLISH] [US-EAST] [100% LAG FREE])
Reporting client version: 512
#43152: 'Panic bunker message is now a config.' by kevinz000 at commit f8777444d6
#42780: 'Automatic deadminning' by ShizCalev at commit aee9417772

Non-core freeform erases all other laws, reset board fails to set some laws up (it resulted in only laws 2 and 3 of asimov getting added to the AI), et cetera. Check logs of round 104102 for the freeform law issue.

### PR DESCRIPTION
**Pull request by timothymtorres**
## About The Pull Request

Fixes #43197

~~@tralezab needs to test this PR before it gets merged.~~ Nevermind I tested it myself.

This formally adds the new AI lawsets from #66636 into the game.  Every lawset can 
 be researched, appears in config, and random spawners.  @tralezab also made several changes to the AI lawsets category that I included below:

Key:
- ❤Harmless (_These shouldn't allow the AI to start butchering people_)
- ⚖ Neutral (_These shouldn't allow the AI to start butchering people without reason_)
- 💀 Harmful (_These will get the shuttle called RIP_)

| Lawset | Now | PR | Unique AI Trait Now | Unique AI Trait PR | 
| ------- | --------- | -------- | -------- | -------- |
| asimov | ❤ |  ❤ | 16% | - |
| asimov++ | ❤ |  ❤ | 6% | 5% |
| hippocratic | ❤ |  ❤ | 6% | 5% |
| paladin 3e | ❤ |  ❤ | 6% | 5% |
| paladin 5e | ❤ |  ❤ | 6% | 5% |
| drone | ❤ |  ❤ | 2% | - |
| united nations | ⚖ | ⚖ | - | - |
| H.O.G.A.N. | ⚖ | ⚖ | 2% | 3% |
| reporter | ⚖ | ⚖ | 2%  | 3% |
| antimov | 💀 | 💀 | - | - |
| balance | 💀 | 💀 | - | - |
| thermodynamic | 💀 | 💀 | - | - |
| SyndieOS | 💀 | 💀 | - | - |
| SpiderOS | 💀 | 💀 | - | - |
| - | - | - | - | - |
| corporation | ⚖ | ❤ | 6% | 5% |
| robocop | ⚖ | ❤ | 6% | 5% |
| maintain | ⚖ | ❤ | 6% | 5% |
| liveandletlive | ⚖| ❤ | 6% | 5% |
| peacekeeper | ⚖ | ❤ | 6% | 5% |
| ten commandments | ⚖ | ❤ | 6% | 5% |
| tyrant | 💀 | ⚖| 2%  | 3% |
| overlord | 💀 | ⚖| 2%  | 3% |
| nutimov | - | ❤ | 2% | 5% |
| Dungeon Master | - | ⚖| 2% | 3% |
| Painter | - | ⚖ | 2%  | 3% |

There is a 75% to select a harmless lawset, and 25% for neutral lawset.

## Why It's Good For The Game

Every AI lawsets can now be properly researched.  They can also appear for the station trait or in a random spawner in AI upload.  The new lawsets (nutimov, painter, DM) now exist in the game.

## Changelog

:cl:
add: Add all AI lawsets can now be researched and have their modules printed
add: Add all AI lawsets to random spawners in AI upload
add: Add advanced AI techweb node
balance: Change AI lawsets to be in different random spawner categories
balance: Change lawsets chance for unique AI station trait
balance: Move some lawsets out of AI techweb node and put into advanced AI node
code: Add documentation for AI lawset code
config: Add every AI lawset to game_options config
config: Rebalance AI lawsets in game_options config
fix: Lawsets overflow to behave correctly 
/:cl:

### CODE REVIEW

File: code/game/objects/items/AI_modules.dm [hunk 1]
```diff
@@ -603,3 +602,21 @@ AI MODULES
 /obj/item/ai_module/core/full/ten_commandments
 	name = "'10 Commandments' Core AI Module"
 	law_id = "ten_commandments"
+
+/********************* Nutimov *********************/
```
Comment (tralezab) [inline code/game/objects/items/AI_modules.dm:606]: no fancy headers

File: code/game/objects/items/AI_modules.dm [hunk 2]
```diff
@@ -603,3 +602,21 @@ AI MODULES
...
+
+/obj/item/ai_module/core/full/nutimov
+	name = "'Nutimov' Core AI Module"
+	law_id = "nutimov"
+
+/******************* Dungeon Master ****************/
```
Comment (tralezab) [inline code/game/objects/items/AI_modules.dm:612]: no fancy headers

File: code/game/objects/items/AI_modules.dm [hunk 3]
```diff
@@ -603,3 +602,21 @@ AI MODULES
...
+
+/obj/item/ai_module/core/full/dungeon_master
+	name = "'Dungeon Master' Core AI Module"
+	law_id = "dungeon_master"
+
+/********************** Painter *******************/
```
Comment (tralezab) [inline code/game/objects/items/AI_modules.dm:618]: no fancy headers

**Review by tralezab — CHANGES_REQUESTED** [review]
lobotomizing the graph from my mind for a moment this looks great. bonus points for removing the rest of the headers on that file

File: code/modules/research/designs/AI_module_designs.dm [hunk 4]
```diff
@@ -153,3 +153,165 @@
 	build_path = /obj/item/ai_module/core/full/custom
 	category = list("AI Modules")
 	departmental_flags = DEPARTMENTAL_FLAG_SCIENCE
+
+/datum/design/board/dungeon_master_module
+	name = "Core Module Design (Dungeon Master)"
+	desc = "Allows for the construction of a Dungeon Master AI Core Module."
+	id = "corporate_module"
```
Comment (magatsuchi) [inline code/modules/research/designs/AI_module_designs.dm:160]: i think you forgot to change this, which is leading to a failed unit test on your checks

**Review by azrael-x86 — CHANGES_REQUESTED** [review]

**Comment by Colovorat** [comment]
💀

**Comment by itseasytosee** [comment]
whats with the table?

**Comment by Redrover1760** [comment]
Mother drone turns AI into a lonely ghost role. Honestly don't know why its a possible roundstart lawset when it's basically a punishment lawset.

File: code/modules/research/designs/AI_module_designs.dm (hunk 4 above)
Comment (timothymtorres) [reply code/modules/research/designs/AI_module_designs.dm]: Fixed with https://github.com/tgstation/tgstation/pull/66854/commits/d70169df6529e124236033a13bad3de4cd119b58 thanks!

File: code/game/objects/items/AI_modules.dm (hunk 1 above)
Comment (timothymtorres) [reply code/game/objects/items/AI_modules.dm]: Removed with https://github.com/tgstation/tgstation/pull/66854/commits/cd8028beb2d16552488fd41e5c64684fd82552b9

File: code/game/objects/items/AI_modules.dm (hunk 2 above)
Comment (timothymtorres) [reply code/game/objects/items/AI_modules.dm]: Removed with https://github.com/tgstation/tgstation/pull/66854/commits/cd8028beb2d16552488fd41e5c64684fd82552b9

File: code/game/objects/items/AI_modules.dm (hunk 3 above)
Comment (timothymtorres) [reply code/game/objects/items/AI_modules.dm]: Removed with https://github.com/tgstation/tgstation/pull/66854/commits/cd8028beb2d16552488fd41e5c64684fd82552b9

**Comment by timothymtorres** [comment]
Alright I moved drone to the harmless lawset.  Also rebalanced the weight chance for unique AI station trait.

**Comment by Timberpoes** [comment]
Don't worry too much about perfectly balanced numbers in the config. Pick sane enough defaults and you're good.

Our config has to be manually updated so the headmins ultimately decide which lawsets they do and do not want and the associated weights. They can (and probably will) ignore the numbers you have here. The same likely applies to downstream too.

Also mother drone is widely considered a grief lawset to punish station AIs. It may be a wise consideration not to enable it by default even if it's harmless, because the harm is in silicon players ghosting or going SSD when they have to experience it.

**Comment by timothymtorres** [comment]
This has been tested and is ready for review @MrStonedOne 

The behavior for overflow laws was pretty broken.  If you added a freeform supplied law and it overflowed, it could replace a core/ion law which didn't make any sense.  Ion/zeroth laws during overflows could replace a supplied law and would be placed as a supplied law.  Now it just deletes a random law and inserts it into the correct category (with a random location in the category)  It should be fixed and consistent now.

Also the metastation snake in the boot CI test failed but that has nothing to do with my code.

**Comment by optimumtact** [comment]
why are you pinging mso, the headadmins should be the ones signing this off

also rulesets are entirely configurable by text, so I'm not even sure what the point of this is.

**Comment by timothymtorres** [comment]
I was under the impression that only the host could approve the config changes.  Good to know that isn't the case for the future.

**Review by MrMelbert — COMMENTED** [review]
only changes the default config so it seems fine

**Comment by optimumtact** [comment]
conflicts

**Pull request merged**